ketWithRadius(Lnet/minecraft/server/level/TicketType;"
             + "Lnet/minecraft/world/level/ChunkPos;I)V"))
     private void cc_onSetDefaultSpawnPos_removeTicketWithRadius(
@@ -100,13 +114,60 @@ public abstract class MixinServerLevel extends MixinLevel implements CubicServer
         if (levelClo instanceof LevelCube levelCube) {
             cc_tickCube(levelCube, randomTickSpeed);
         } else {
-            // TODO (P2) chunk ticking for anything that still needs to happen on chunks (probably just the forge event for mods?)
+            // Columns remain authoritative for precipitation and weather. Cubes own random block/fluid ticks.
+            ((ServerLevel) (Object) this).tickChunk((LevelChunk) levelClo, 0);
         }
     }
 
     @AddMethodToSets(containers = ChunkToCubeSet.ServerLevel_redirects.class, method = "tickChunk(Lnet/minecraft/world/level/chunk/LevelChunk;I)V")
     public void cc_tickCube(LevelCube levelCube, int randomTickSpeed) {
-        // TODO (P2) cube ticking
+        if (randomTickSpeed <= 0) {
+            return;
+        }
+
+        ServerLevel serverLevel = (ServerLevel) (Object) this;
+        CubePos cubePos = levelCube.cc_getCubePos();
+        LevelChunkSection[] sections = levelCube.getSections();
+        ProfilerFiller profiler = Profiler.get();
+        profiler.push("tickBlocks");
+        try {
+            for (int localSectionY = 0; localSectionY < CubicConstants.DIAMETER_IN_SECTIONS; localSectionY++) {
+                for (int localSectionZ = 0; localSectionZ < CubicConstants.DIAMETER_IN_SECTIONS; localSectionZ++) {
+                    for (int localSectionX = 0; localSectionX < CubicConstants.DIAMETER_IN_SECTIONS; localSectionX++) {
+                        int sectionIndex = Coords.sectionToIndex(localSectionX, localSectionY, localSectionZ);
+                        LevelChunkSection section = sections[sectionIndex];
+                        if (!section.isRandomlyTicking()) {
+                            continue;
+                        }
+
+                        int minBlockX = SectionPos.sectionToBlockCoord(Coords.cubeToSection(cubePos.getX(), localSectionX));
+                        int minBlockY = SectionPos.sectionToBlockCoord(Coords.cubeToSection(cubePos.getY(), localSectionY));
+                        int minBlockZ = SectionPos.sectionToBlockCoord(Coords.cubeToSection(cubePos.getZ(), localSectionZ));
+                        for (int tickIndex = 0; tickIndex < randomTickSpeed; tickIndex++) {
+                            BlockPos blockPos = serverLevel.getBlockRandomPos(minBlockX, minBlockY, minBlockZ, SectionPos.SECTION_MASK);
+                            profiler.push("randomTick");
+                            try {
+                                BlockState blockState = section.getBlockState(
+                                        blockPos.getX() - minBlockX, blockPos.getY() - minBlockY, blockPos.getZ() - minBlockZ
+                                );
+                                if (blockState.isRandomlyTicking()) {
+                                    blockState.randomTick(serverLevel, blockPos, serverLevel.random);
+                                }
+
+                                FluidState fluidState = blockState.getFluidState();
+                                if (fluidState.isRandomlyTicking()) {
+                                    fluidState.randomTick(serverLevel, blockPos, serverLevel.random);
+                                }
+                            } finally {
+                                profiler.pop();
+                            }
+                        }
+                    }
+                }
+            }
+        } finally {
+            profiler.pop();
+        }
     }
 
     // TODO (P2) waitForChunkAndEntities
diff --git a/src/main/java/io/github/opencubicchunks/cubicchunks/server/level/CubicServerLevel.java b/src/main/java/io/github/opencubicchunks/cubicchunks/server/level/CubicServerLevel.java
index